tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -26,53 +26,94 @@ tyToStr(const Type *T) {
 
 PointerVariableConstraint::PointerVariableConstraint(DeclaratorDecl *D,
   uint32_t &K, Constraints &CS, const ASTContext &C) :
-  PointerVariableConstraint(D->getType(), K, D->getName(), CS, C) { }
+  PointerVariableConstraint(D->getType(), K, D, D->getName(), CS, C) { }
 
 PointerVariableConstraint::PointerVariableConstraint(const QualType &QT, uint32_t &K,
-	std::string N, Constraints &CS, const ASTContext &C) : 
+	DeclaratorDecl *D, std::string N, Constraints &CS, const ASTContext &C) : 
 	ConstraintVariable(ConstraintVariable::PointerVariable, 
-					   tyToStr(QT.getTypePtr())),FV(nullptr)
-{ 
+					   tyToStr(QT.getTypePtr()),N),FV(nullptr)
+{   
 	QualType QTy = QT;
 	const Type *Ty = QTy.getTypePtr();
+  // If the type is a decayed type, then maybe this is the result of 
+  // decaying an array to a pointer. If the original type is some 
+  // kind of array type, we want to use that instead. 
+  if (const DecayedType *DC = dyn_cast<DecayedType>(Ty)) {
+    QualType QTytmp = DC->getOriginalType();
+    const Type * Tytmp = QTytmp.getTypePtr();
```

Comment:
Probably, I'll try to make it cleaner. `isArrayType` won't work on `Ty` though, in the case of an array like `int arr[]` because it becomes a `DecayedType` that contains both an original type (which *is* an `ArrayType`) and a "pointer type."